sdk/storage/azure-storage-common/inc/azure/storage/common/crypt.hpp

```diff
@@ -7,32 +7,21 @@
 #include <string>
 #include <vector>
 
+#include "azure/core/cryptography/hash.hpp"
 #include <azure/core/base64.hpp>
 
 namespace Azure { namespace Storage {
 
-  class Crc64 {
+  class Crc64Hash : public Azure::Core::Cryptography::Hash {
   public:
-    void Update(const uint8_t* data, std::size_t length);
-    void Concatenate(const Crc64& other);
-
-    std::vector<uint8_t> Digest() const;
-
-    static std::vector<uint8_t> Hash(const uint8_t* data, std::size_t length)
-    {
-      Crc64 instance;
-      instance.Update(data, length);
-      return instance.Digest();
-    }
-
-    static std::vector<uint8_t> Hash(const std::string& data)
-    {
-      return Hash(reinterpret_cast<const uint8_t*>(data.data()), data.length());
-    }
+    void Concatenate(const Crc64Hash& other);
 
```

Comment:
need override destructor

Reply:
What will we do in the destructor? It wasn't there in the previous implementation.

Reply:
nothing, just 

```C++
Crc64Hash() override = default;
```

to enforce the base class has a virtual destructor.

Reply:
Seems odd to do. The base class already has a virtual destructor, why would the concrete derived class need to enforce such a thing?

Reply:
Let's say some customer cares a lot about data integrity and always wants to use Md5 or Crc64 or whatever checksum servicer supports. He also wants to be able to switch different hash algorithm with an option, possible value of which can be MD5, CRC64, and others. To avoid duplicate logic, he will use a pointer to the hash base class `HashBase*` and instantiate a concrete hash instance `Md5Hash` or `Crc64Hash` and assign the point to `HashBase*`. Now he can just call `HashBase->Append` to append data and `HashBase->Final` to get final hash value, without caring about the hash algorithm. This helps a lot in reducing code redundancy. Once the operation is done, he will deallocate the hash instance by `delete HashBase`. If the destructor of `HashBase` is not virtual, it will cause potential resource leak because the destructor of concrete hash class isn't executed. Now to avoid this, we will use `override` on the destructor of concrete class. It's always a good habit if some object is possible to be deleted via base pointer.

> The base class already has a virtual destructor, why would the concrete derived class need to enforce such a thing? What's the benefit?

same question goes for [OnFinal](https://github.com/Azure/azure-sdk-for-cpp/blob/2d4d766c6d454bfbd2a3627c5a3d8a4d81196480/sdk/core/azure-core/inc/azure/core/cryptography/hash.hpp#L136) and [OnAppend](https://github.com/Azure/azure-sdk-for-cpp/blob/2d4d766c6d454bfbd2a3627c5a3d8a4d81196480/sdk/core/azure-core/inc/azure/core/cryptography/hash.hpp#L146). We don't need `override` to be grammatically correct.

Reply:
> If the destructor of `HashBase` is not virtual, it will cause potential resource leak because the destructor of concrete hash class isn't executed.

But it is virtual, so why safeguard against something that is already true?

> same question goes for [OnFinal](https://github.com/Azure/azure-sdk-for-cpp/blob/2d4d766c6d454bfbd2a3627c5a3d8a4d81196480/sdk/core/azure-core/inc/azure/core/cryptography/hash.hpp#L136) and [OnAppend](https://github.com/Azure/azure-sdk-for-cpp/blob/2d4d766c6d454bfbd2a3627c5a3d8a4d81196480/sdk/core/azure-core/inc/azure/core/cryptography/hash.hpp#L146). We don't need `override` to be grammatically correct. I just think it's a good habit

But in this case, we are actually overriding the behavior/logic of those functions.

The `HashBase` already has an implementation of the virtual destructor as `default`, and if `Crc64Hash` is not changing the implementation or doing something meaningful, why override it prematurely? I agree that **IF** the destructor needs to be overridden, we'd use the `override` keyword. I am asking what's the benefit of overriding it today?

All that said, I am fine with adding it.

Reply:
> But in this case, we are actually overriding the behavior/logic of those functions.

Without `override`, you're also overriding the behavior.

> The HashBase already has an implementation of the virtual destructor as default, and if Crc64Hash is not changing the implementation or doing something meaningful, why override it prematurely?

As you say, it's just a safeguard. Who's to say nobody will change the impl/interface of `HashBase` and forget to make the destructor virtual. If we use `override` in derived destructor, the compiler can remind us.

Reply:
> Without `override`, you're also overriding the behavior.

But currently I am not providing a destructor for `Crc64Hash` at all. How am I overriding anything? I understand **IF** I did provide one, the `override` keyword is helpful, but if I don't add a destructor at all, what's the issue?

Are you suggesting I **add** a destructor?